### ISSUE TITLE
Default value of year should be converted?

### ISSUE THREAD

**Issue by jackysp** [pingcap/tidb#10191]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
create table t2(
y year NOT NULL DEFAULT '0001',
d date NOT NULL DEFAULT '0000-01-01',
t time NOT NULL DEFAULT '00:00:00',
ts timestamp NOT NULL DEFAULT '2038-01-19 03:14:07.999999',
dt datetime NOT NULL DEFAULT '2038-01-19 03:14:07.0000000'
)ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
SHOW CREATE TABLE;
```

2. What did you expect to see?
```
create table t2(
y year NOT NULL DEFAULT '0001',
d date NOT NULL DEFAULT '0000-01-01',
t time NOT NULL DEFAULT '00:00:00',
ts timestamp NOT NULL DEFAULT '2038-01-19 03:14:07.999999',
dt datetime NOT NULL DEFAULT '2038-01-19 03:14:07.0000000'
)ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
```

3. What did you see instead?
```
create table t2(
y year NOT NULL DEFAULT '2001',
d date NOT NULL DEFAULT '0000-01-01',
t time NOT NULL DEFAULT '00:00:00',
ts timestamp NOT NULL DEFAULT '2038-01-19 03:14:07.999999',
dt datetime NOT NULL DEFAULT '2038-01-19 03:14:07.0000000'
)ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

3.0.0-beta

**Comment by erjiaqing** [comment]
According to https://dev.mysql.com/doc/refman/5.7/en/year.html, 0001 is not a valid year.

And TiDB's behavior is same as MySQL

**Comment by wjhuang2016** [comment]
@erjiaqing 
I had the same feeling with you at first. And then I realised that MySQL takes `0001` as `1`.

**Comment by nullnotnil** [comment]
In master, the column `ts` can not be created in TiDB or MySQL (this is correct behavior, since the value is out of range). But I can reproduce the default value of year not being converted to 2001 as in MySQL:

```sql
drop table if exists t2;
create table t2 (
y year NOT NULL DEFAULT '0001',
d date NOT NULL DEFAULT '0000-01-01',
t time NOT NULL DEFAULT '00:00:00',
dt datetime NOT NULL DEFAULT '2038-01-19 03:14:07.0000000'
);
show create table t2\G
SELECT tidb_version()\G

..

mysql> show create table t2\G
*************************** 1. row ***************************
       Table: t2
Create Table: CREATE TABLE `t2` (
  `y` year(4) NOT NULL DEFAULT '0001',
  `d` date NOT NULL DEFAULT '0000-01-01',
  `t` time NOT NULL DEFAULT '00:00:00',
  `dt` datetime NOT NULL DEFAULT '2038-01-19 03:14:07'
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
1 row in set (0.00 sec)

mysql> SELECT tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v4.0.0-beta.2-750-g8a661044c
Edition: Community
Git Commit Hash: 8a661044cedf8daad1de4fbf79a390962b6f6c3b
Git Branch: master
UTC Build Time: 2020-07-10 10:52:37
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)
```

**Comment by bb7133** [comment]
Looks that this bug has been fixed in TiDB v3.0/4.0/master.
